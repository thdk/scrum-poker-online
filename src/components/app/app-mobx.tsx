import * as React from 'react';
import { observer } from 'mobx-react-lite';

import { App } from './app';
import { useAuthStore } from '../../modules/auth/use-auth-store';

const AppMobx = () => {
  const auth = useAuthStore();

  if (!auth.isAuthInitialised || !auth.collection.isFetched) {
    return <></>;
  }

  return <App />;
};

export default observer(AppMobx);
