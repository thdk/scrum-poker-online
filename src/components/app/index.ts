export { default } from './app-mobx';
