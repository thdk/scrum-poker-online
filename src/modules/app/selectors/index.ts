export * from './player';
export * from './view';
