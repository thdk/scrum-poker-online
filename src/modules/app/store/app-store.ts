import { PlayerStore } from "../../player/store";
import { observable, computed, action } from "mobx";
import Cookie from "mobx-cookie";
import { ViewStore } from "../../view/store";
import firebase from "firebase/app";

export type PlayCard = 0 | 1 | 2 | 3 | 5 | 8 | 13 | 20 | 40 | 100 | "coffee" | "?";

export const sortCards = (a: PlayCard, b: PlayCard) => {
	if (a === b) return 0;

	if (a === "coffee")
		return 1;

	if (a === "?")
		return -1;

	return a < b ? 1 : -1;
}

export class AppStore {
	public playerStore: PlayerStore;
	public viewStore: ViewStore;

	@observable
	private sessionCodeCookie = new Cookie("session-code");

	constructor({
		firestore,
	}: {
		firestore: firebase.firestore.Firestore;
	}) {
		const urlParams = new URLSearchParams(window.location.search);

		const sessionFromUrl = urlParams.get("session");
		if (sessionFromUrl) {
			this.setSessionCode(sessionFromUrl);
		}

		this.playerStore = new PlayerStore(
			this,
			{
				firestore,
			},
		);
		this.viewStore = new ViewStore();
	}


	public get cards(): PlayCard[] {
		return [
			0, 1, 2, 3, 5, 8, 13, 20, 40, 100, "coffee", "?",
		];
	}

	@computed
	public get sessionCode() {
		return this.sessionCodeCookie.value;
	}

	@action
	public setSessionCode(value: string | undefined) {
		if (value) {
			this.sessionCodeCookie.set(value, { expires: 2 });
		} else {
			this.sessionCodeCookie.remove();
		}
	}
}
