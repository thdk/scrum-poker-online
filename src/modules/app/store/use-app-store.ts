import firebase from 'firebase/app';
import React from 'react';
import { AppStore } from '.';

const createStore = () => {
  const store = new AppStore({
    firestore: firebase.firestore(),
    auth: firebase.auth(),
  });

  // for development purposes only
  if (process.env.NODE_ENV !== 'production') {
    (window as any).store = store;
  }

  const urlParams = new URLSearchParams(window.location.search);
  store.viewStore.sessionFromUrl = urlParams.get('session') || undefined;

  return store;
};

export type Store = ReturnType<typeof createStore>;

export const StoreContext = React.createContext(createStore());

export const useStore = () => {
  const store = React.useContext(StoreContext);
  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }
  return store;
};
