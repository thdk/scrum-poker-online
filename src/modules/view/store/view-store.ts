import { observable } from 'mobx';

export class ViewStore {
  @observable
  public isEditMode = false;
}
