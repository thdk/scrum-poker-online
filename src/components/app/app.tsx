import React from 'react';
import './app.css';
import { WelcomeScreen } from '../../screens/welcome';
import GameScreen from '../../screens/game';
import Header from '../header';

export type AppProps = {
  screen: 'game' | 'settings';
};

export const App = (props: AppProps) => {
  const { screen } = props;

  return (
    <div>
      <Header />
      {
        screen === 'game'
          ? <GameScreen />
          : <WelcomeScreen />
      }
    </div>
  );
};
