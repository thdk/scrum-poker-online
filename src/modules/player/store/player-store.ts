import { Collection, CollectionReference, Doc } from "firestorable";
import firebase from "firebase/app";
import { computed, observable, reaction, action } from "mobx";
import MobxCookie from "mobx-cookie";

import { AppStore, sortCards } from "../../app/store";
import { IPlayer, IPlayerData, PlayerType } from "../types";
import { serializePlayer } from "../serialization/serializer";
import { deserializePlayer } from "../serialization/deserializer";
import { getSessionCode } from "../../app/selectors";

export class PlayerStore {
	public collection: Collection<IPlayer, IPlayerData>;

	@observable
	private playerIdCookie = new MobxCookie("player-id");

	@observable.ref
	private playerDoc?: Doc<IPlayer>;

	private readonly appStore: AppStore;

	constructor(
		appStore: AppStore,
		{
			firestore,
		}: {
			firestore: firebase.firestore.Firestore,
		}
	) {
		this.appStore = appStore;

		const createQuery = () => {
			const session = getSessionCode(appStore);

			return session ?
				(ref: CollectionReference) =>
					ref.where("session", "==", session)
						.where("type", "==", PlayerType.player)
				: null;
		};

		this.collection = new Collection<IPlayer, IPlayerData>(
			firestore,
			"players",
			{
				query: createQuery(),
				serialize: serializePlayer,
				deserialize: deserializePlayer,
			},
		);

		reaction(() => getSessionCode(appStore), () => {
			this.collection.query = createQuery();
		});

		reaction(
			() => this.playerId,
			id => {
				id && this.getPlayerAsync(id);
			},
			{
				fireImmediately: true,
			},
		);
	}

	@computed
	private get playerId() {
		return this.playerIdCookie.value;
	}

	@action
	private setPlayerId(value: string | undefined) {
		if (value) {
			this.playerIdCookie.set(value);
		} else {
			this.playerIdCookie.remove();
		}
	}

	@computed
	public get allPlayers() {
		return this.collection.docs;
	}

	@computed get allPlayersSorted() {
		if (this.isWaitingForplayer)
			return this.allPlayers;

		return this.allPlayers.sort((a, b) => sortCards(a.data!.value!, b.data!.value!));
	}

	@computed
	public get player() {
		const playerDoc = this.playerDoc;

		return playerDoc && playerDoc.data;
	}

	@computed
	public get isWaitingForplayer() {
		return this.allPlayers.some(p => p.data!.value === undefined);
	}

	public savePlayer(data: IPlayer) {
		this.player
			? this.updatePlayerAsync(data)
			: this.createPlayerAsync(data);
	}

	private getPlayerAsync(id: string) {
		this.collection.getAsync(id, { watch: true })
			.then(playerDoc => {
				if (playerDoc.data!.session)
					this.playerDoc = playerDoc;
				this.validatePlayerSession();
			}, () => {
				this.setPlayerId(undefined);
			});
	}

	private validatePlayerSession() {
		const currentSessionCode = getSessionCode(this.appStore)
		if (this.player && (
			this.player.value !== undefined // Reset any value from the previous session
			|| (currentSessionCode && this.player.session !== currentSessionCode)) // Reset the players session to the current session
		) {
			this.updatePlayerAsync({
				session: currentSessionCode,
				value: undefined
			});
		}
	}

	public createPlayerAsync(player: IPlayer) {
		this.appStore.setSessionCode(player.session);

		return this.collection.addAsync(player)
			.then(this.setPlayerId.bind(this));
	}

	public updatePlayerAsync(player: Partial<IPlayer>) {
		if (player.session) {
			this.appStore.setSessionCode(player.session);
		}

		return (this.player && this.playerId)
			? this.collection.updateAsync(player, this.playerId)
			: Promise.reject(new Error("No player set"));
	}

	public restart() {
		this.collection.updateAsync(
			{
				value: undefined,
			},
			...this.collection.docs.map(p => p.id),
		)
	}
}
