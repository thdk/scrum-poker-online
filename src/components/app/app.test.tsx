import React from 'react';
import { render } from '@testing-library/react';
import App from '.';

import {
  initializeAdminApp,
} from "@firebase/rules-unit-testing";
import { AppStore } from '../../modules/app/store';

const app = initializeAdminApp({
  projectId: "app-test",
});

afterAll(() => {
  app.delete();
});

const mockStore = new AppStore({
  firestore: app.firestore(),
});

jest.mock("../../hooks", () => ({
  useStore: () => mockStore,
}));

test('renders', () => {
  const { asFragment } = render(<App />);
  expect(asFragment()).toMatchSnapshot();
});
