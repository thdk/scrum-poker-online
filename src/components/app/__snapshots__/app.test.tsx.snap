// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div>
    <div
      class="header"
    >
      <div
        class="header-tab active"
      >
        GAME
      </div>
      <div
        class="header-tab"
      >
        SETTINGS
      </div>
    </div>
    <div
      class="screen welcome"
    >
      <div
        class="player-form"
      >
        <div
          class="player-form-fields content"
        >
          <div
            class="form-field form-field-checkbox"
          >
            <div
              class="form-field-label"
            >
              Tv mode
            </div>
            <div
              class="form-field-input"
            >
              <input
                type="checkbox"
              />
            </div>
          </div>
          <div
            class="form-field form-field-text"
          >
            <div
              class="form-field-label"
            >
              Session code
            </div>
            <div
              class="form-field-input"
            >
              <input
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="form-field form-field-text"
          >
            <div
              class="form-field-label"
            >
              Player name
            </div>
            <div
              class="form-field-input"
            >
              <input
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="form-field form-field-button"
          >
            <div
              class="form-field-input"
            >
              <input
                type="button"
                value="Save"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
